### ISSUE TITLE
File > Open = is very slow to load, sometimes windows explorer locks up and becomes (Not Responding)

### ISSUE THREAD

**Issue by ImSpecial** [SourMesen/Mesen#562]
The issue here is that when this [**Windows Explorer window**](https://i.imgur.com/omFK0Fb.png) pops up, it is very slow to load the list of ROMs inside the folder, and sometimes just hangs for a full second or two and becomes (Not Responding), at first I thought it might be the large number of files inside the folder, but other NES emulators such as Nestopia and NesterJJJ, do not have this problem, and other emulators that also use the same kind of File > Open window such as SNES9x and GENs which have equally sized folders give or take are much quicker to populate their list then Mesen does.

So this seems to be a Mesen issue with me, does Mesen parse or check what's inside each zip files on load or something? That's the only thing I can think of that would cause such a slow load, it's rather painful specially when I go inside my Translation folder when back out, it needs to reload everything again, basically any action requires a reload.

I'm hopeful that something can be figured out about this issue.

**Comment by SourMesen** [comment]
Mesen just uses the standard .NET open file dialogs and doesn't do anything special with it (other than setting the available extensions and the folder in which the dialog opens up).  If it's freezing up, it's almost guaranteed that this is caused by .NET's code or Windows' own code for the dialog.

This kind of slowdown can sometimes be caused by the shortcuts/pinned elements you have on the left side of the window. Though I'm not sure why that would only be affecting Mesen.

You can try [Visual NES](https://www.mesen.ca/VisualNes.zip) to see if it behaves the same as Mesen or if it's different (but there is no reason why either of them would behave differently, since they're both essentially doing the same thing)

**Comment by ImSpecial** [comment]
Visual NES only reads .nes files and does not pick up any of my .zip files, which they all are. So I can't tell, I'll try and make some gifs of my issue for comparisons, I really know how to make gifs though so we'll see how this goes....

**Comment by SourMesen** [comment]
Yea, that definitely looks odd, but it's also most likely 100% out of my hands - the open file dialog is entirely managed by Windows' code itself - I couldn't even begin to guess as to why it's somehow slower on Mesen :\

**Comment by SourMesen** [comment]
Some random guesses:
-AV software is slowing down the operation for Mesen's process only (maybe because it's a .NET app, etc.)
-WinRAR is causing something to happen in the background for files associated with it and it's for some reason only being slow when done inside Mesen

So you could try removing WinRAR's file associations to see if that changes anything and/or disabling any AV software you might have active.

**Comment by ImSpecial** [comment]
I just use Windows Defender, but giving Mesen an exclude just now doesn't seem to help any, and if WinRaR was the cause, it would I think have effects on other similar emulators/programs that use the same kind of open dialog, It's only been Mesen from what I can tell, googling and I find some similar issues that refer to old shortcuts and old file associations like you said... but I'm pretty sure those fixes don't relate to me, it's weird for sure. Thanks for trying anyway, If I find something via Google or whatever I'll be sure to update this thread.

**Comment by ImSpecial** [comment]
Not sure if you'll see this but just want to let you know that I grabbed the latest dev build to mess around with, DevWin-0.9.7.88 previously been using Win-0.9.7 and there is no slow down anymore when open file dialog opens... it's still there with 0.9.7 so not a Windows fix, [I'm guessing it was this you did?](https://github.com/SourMesen/Mesen/commit/295a68f43dd235ceab889e51e0168d095b8eb641)

Anyway thanks!

**Comment by SourMesen** [comment]
Actually, that particular commit has no link to it - like I said, there is no code in Mesen that is related to displaying that open file dialog, so I actually have no idea why it would be fixed.  Could you try older dev builds to see when the problem disappears? (They're available in the history tab: https://ci.appveyor.com/project/Sour/mesen/history)  As is, I don't know what could have possibly fixed it, and I'm not convinced it really is "fixed", either - might just be that your PC is not reacting to the process the same way because it's a different binary, etc.

The next release will probably be sometime this month, but I don't have any specific date planned, mostly a case of getting through a number of things I want to get done before the next release.

**Comment by ImSpecial** [comment]
You're right to be not convinced I spoke to soon. */sigh

But with some testing I was able to find some stuff out that tells me the cause, no fix for this issue unfortunately.

First a little about my setup, my [Drives](https://i.imgur.com/3A5Q788.png) are setup in a way like this, now this is what I've found, when I downloaded your latest dev build it was downloaded to my "download folder" which is on Drive (D:), that's where I ran it, and the open dialog was lightning quick so I thought things were fixed, but after moving that dev build of Mesen to my Drive (E:) folder, things were slow again, I can reproduce this super easily just by copy/pasting my 0.9.7 Mesen folder to anywhere on my (D:) Drive, run it from there and it's good, run it from another drive that's not the same as the roms, and it's slow again...

I went through all my [emulators](https://i.imgur.com/wqWWheM.png) as tests, and only Mesen and BizHawk have this issue, the rest are lightning quick, and all sit on (E:) pointing to their (D:) counterpart. Perhaps Mesen/BizHawk share something in common like written in the same language? Whereas the other emulators are different in that regard, i dunno.

BizHawk has the same fix as well, copy this emulator folder to my (D:), run it and it'll open the "open dialog instantly", but if it sits on my (E:) Drive it'll be slow again, which I find strange since my (E:) drive is a good SSD, (850 EVO 250GB) so you'd think things would be fast, SSD's are of good health, TRIM, etc. HHD is good health and defragged perfectly, they're also all internal drives, not external plug in/plug out kind, so I don't think this is a hardware issue.

I've tried "tricking" this with Symlinks have the emulator point to what it was thinking is the same drive as it's on, but it's still slow...

In Conclusion/TL;DR: if Mesen (or BizHawk) go A to B, things are slow, if B to B, things are fast. Probably a Windows thing.

**Comment by SourMesen** [comment]
Thanks for the update - of your list, only Mesen & Bizhawk use a UI built in .NET.
Still though, .NET itself is using Windows' standard file dialog, so it should work the same way.  It's possible that part of the .NET-side code is causing the slowdown somehow, or something else on your computer is causing it to slow down, only when the file dialog is opened within a .NET process.  At the very least, on my end Mesen is on my C: (SSD) and games and the like are on a regular HDD (D:), and I'm not seeing any slowdown.  Maybe the WinRar file association is causing the slowdown somehow? I only have 7zip installed, and it's not associated with 7z files, either.

**Comment by ImSpecial** [comment]
Not sure about the WinRar association part, files are all just standard .zip files, I just have WinRar set as my default "open with" program, if WinRar is the cause it's weird how there are no issues with other emulators (besides Mesen/Biz) and how it's not an issue when they are on the same Drive.

**Comment by SourMesen** [comment]
I mostly meant it could be an interaction between the WinRar file association (or something else) and .NET, since I don't have the issue with a very similar setup to your own (and no issues with both a 2000+ zip files folder or 2000+ 7z files folder)

**Comment by ImSpecial** [comment]
Alright, well if I come across anything in the future about this, I'll keep you updated.
